// @remotion/studio/dist/components/Timeline/EasingEditorModal.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EasingEditorModal = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const modals_1 = require("../../state/modals");
const Button_1 = require("../Button");
const layout_1 = require("../layout");
const ModalButton_1 = require("../ModalButton");
const ModalFooter_1 = require("../ModalFooter");
const ModalHeader_1 = require("../ModalHeader");
const DismissableModal_1 = require("../NewComposition/DismissableModal");
const InputDragger_1 = require("../NewComposition/InputDragger");
const update_selected_easing_1 = require("./update-selected-easing");
const SVG_WIDTH = 560;
const SVG_HEIGHT = 320;
const PLOT_LEFT = 42;
const PLOT_TOP = 8;
const PLOT_WIDTH = 500;
const PLOT_HEIGHT = 304;
const Y_MIN = -2;
const Y_MAX = 3;
const LINEAR_BEZIER = [0.25, 0.25, 0.75, 0.75];
const container = {
    width: 600,
};
const coordinatesGrid = {
    display: 'grid',
    gridTemplateColumns: 'repeat(4, minmax(0, 1fr))',
    gap: 10,
    marginTop: 12,
    padding: '0 16px 16px',
};
const coordinateRow = {
    gap: 4,
};
const coordinateLabel = {
    fontSize: 13,
    color: 'rgba(255, 255, 255, 0.72)',
    paddingLeft: 6,
};
const coordinateInputWrapper = {
    height: 34,
    display: 'flex',
    alignItems: 'center',
};
const numberInputStyle = {
    backgroundColor: colors_1.INPUT_BACKGROUND,
    borderRadius: 4,
    width: '100%',
};
const svgStyle = {
    display: 'block',
    width: '100%',
};
const hiddenSubmit = {
    display: 'none',
};
const clamp = (value, min, max) => {
    return Math.min(max, Math.max(min, value));
};
const sanitizeBezier = (bezier) => [
    clamp(bezier[0], 0, 1),
    clamp(bezier[1], Y_MIN, Y_MAX),
    clamp(bezier[2], 0, 1),
    clamp(bezier[3], Y_MIN, Y_MAX),
];
const easingToBezier = (easing) => {
    return easing === 'linear' ? LINEAR_BEZIER : sanitizeBezier(easing);
};
const roundCoordinate = (value) => Math.round(value * 10000) / 10000;
const serializeBezier = (bezier) => {
    const rounded = sanitizeBezier(bezier).map(roundCoordinate);
    if (rounded[0] === rounded[1] && rounded[2] === rounded[3]) {
        return 'linear';
    }
    return rounded;
};
const formatNumber = (value) => {
    const numericValue = Number(value);
    if (!Number.isFinite(numericValue)) {
        return String(value);
    }
    return String(roundCoordinate(numericValue));
};
const xToSvg = (value) => PLOT_LEFT + value * PLOT_WIDTH;
const yToSvg = (value) => PLOT_TOP + ((Y_MAX - value) / (Y_MAX - Y_MIN)) * PLOT_HEIGHT;
const pointFromBezier = (bezier, handle) => {
    const x = handle === 0 ? bezier[0] : bezier[2];
    const y = handle === 0 ? bezier[1] : bezier[3];
    return { x: xToSvg(x), y: yToSvg(y) };
};
const startPoint = { x: xToSvg(0), y: yToSvg(0) };
const endPoint = { x: xToSvg(1), y: yToSvg(1) };
const EasingEditorModal = ({ state }) => {
    const { setSelectedModal } = (0, react_1.useContext)(modals_1.ModalsContext);
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const sequencesRef = (0, react_1.useContext)(remotion_1.Internals.SequenceManagerRefContext);
    const propStatusesRef = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesRefContext);
    const { setPropStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModeSettersContext);
    const { overrideIdToNodePathMappings } = (0, react_1.useContext)(remotion_1.Internals.OverrideIdsToNodePathsGettersContext);
    const svgRef = (0, react_1.useRef)(null);
    const [bezier, setBezier] = (0, react_1.useState)(() => easingToBezier(state.initialEasing));
    const [saving, setSaving] = (0, react_1.useState)(false);
    const [activeHandle, setActiveHandle] = (0, react_1.useState)(null);
    const close = (0, react_1.useCallback)(() => {
        setSelectedModal(null);
    }, [setSelectedModal]);
    const setCoordinate = (0, react_1.useCallback)((handle, coordinate, value) => {
        setBezier((previous) => {
            const next = [...previous];
            const index = handle === 0
                ? coordinate === 'x'
                    ? 0
                    : 1
                : coordinate === 'x'
                    ? 2
                    : 3;
            next[index] =
                coordinate === 'x' ? clamp(value, 0, 1) : clamp(value, Y_MIN, Y_MAX);
            return next;
        });
    }, []);
    const getValueFromPointer = (0, react_1.useCallback)((event) => {
        const svg = svgRef.current;
        if (!svg) {
            return null;
        }
        const rect = svg.getBoundingClientRect();
        const svgX = ((event.clientX - rect.left) / rect.width) * SVG_WIDTH;
        const svgY = ((event.clientY - rect.top) / rect.height) * SVG_HEIGHT;
        const x = clamp((svgX - PLOT_LEFT) / PLOT_WIDTH, 0, 1);
        const y = clamp(Y_MAX - ((svgY - PLOT_TOP) / PLOT_HEIGHT) * (Y_MAX - Y_MIN), Y_MIN, Y_MAX);
        return { x, y };
    }, []);
    const updateHandleFromPointer = (0, react_1.useCallback)((handle, event) => {
        const value = getValueFromPointer(event);
        if (!value) {
            return;
        }
        setBezier((previous) => {
            const next = [...previous];
            if (handle === 0) {
                next[0] = value.x;
                next[1] = value.y;
            }
            else {
                next[2] = value.x;
                next[3] = value.y;
            }
            return next;
        });
    }, [getValueFromPointer]);
    (0, react_1.useEffect)(() => {
        if (activeHandle === null) {
            return;
        }
        const onPointerMove = (event) => {
            updateHandleFromPointer(activeHandle, event);
        };
        const onPointerUp = () => {
            setActiveHandle(null);
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp, { once: true });
        return () => {
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
        };
    }, [activeHandle, updateHandleFromPointer]);
    const onHandlePointerDown = (0, react_1.useCallback)((handle, event) => {
        event.preventDefault();
        event.stopPropagation();
        setActiveHandle(handle);
        updateHandleFromPointer(handle, event);
    }, [updateHandleFromPointer]);
    const onSave = (0, react_1.useCallback)(() => {
        if (previewServerState.type !== 'connected' || saving) {
            return;
        }
        setSaving(true);
        const promise = (0, update_selected_easing_1.updateSelectedTimelineEasings)({
            selections: state.selections,
            sequences: sequencesRef.current,
            overrideIdsToNodePaths: overrideIdToNodePathMappings,
            propStatuses: propStatusesRef.current,
            setPropStatuses,
            clientId: previewServerState.clientId,
            easing: serializeBezier(bezier),
        });
        if (promise === null) {
            setSaving(false);
            return;
        }
        promise.catch(() => undefined);
        close();
    }, [
        bezier,
        close,
        overrideIdToNodePathMappings,
        previewServerState,
        propStatusesRef,
        saving,
        sequencesRef,
        setPropStatuses,
        state.selections,
    ]);
    const onSubmit = (0, react_1.useCallback)((event) => {
        event.preventDefault();
        onSave();
    }, [onSave]);
    const path = (0, react_1.useMemo)(() => {
        const curveFirstHandle = pointFromBezier(bezier, 0);
        const curveSecondHandle = pointFromBezier(bezier, 1);
        return `M ${startPoint.x} ${startPoint.y} C ${curveFirstHandle.x} ${curveFirstHandle.y}, ${curveSecondHandle.x} ${curveSecondHandle.y}, ${endPoint.x} ${endPoint.y}`;
    }, [bezier]);
    const firstHandle = pointFromBezier(bezier, 0);
    const secondHandle = pointFromBezier(bezier, 1);
    const yZero = yToSvg(0);
    const yOne = yToSvg(1);
    const saveDisabled = saving || previewServerState.type !== 'connected';
    return (jsx_runtime_1.jsxs(DismissableModal_1.DismissableModal, { children: [
            jsx_runtime_1.jsx(ModalHeader_1.ModalHeader, { title: "Edit easing" }), jsx_runtime_1.jsxs("form", { onSubmit: onSubmit, children: [
                    jsx_runtime_1.jsxs("div", { style: container, children: [
                            jsx_runtime_1.jsxs("svg", { ref: svgRef, width: SVG_WIDTH, height: SVG_HEIGHT, viewBox: `0 0 ${SVG_WIDTH} ${SVG_HEIGHT}`, style: svgStyle, "aria-label": "Bezier curve editor", children: [
                                    jsx_runtime_1.jsx("line", { x1: PLOT_LEFT, y1: yZero, x2: PLOT_LEFT + PLOT_WIDTH, y2: yZero, stroke: colors_1.INPUT_BORDER_COLOR_HOVERED, strokeWidth: 1 }), jsx_runtime_1.jsx("line", { x1: PLOT_LEFT, y1: yOne, x2: PLOT_LEFT + PLOT_WIDTH, y2: yOne, stroke: colors_1.INPUT_BORDER_COLOR_HOVERED, strokeWidth: 1 }), jsx_runtime_1.jsx("line", { x1: startPoint.x, y1: startPoint.y, x2: firstHandle.x, y2: firstHandle.y, stroke: "rgba(255, 255, 255, 0.35)", strokeWidth: 1 }), jsx_runtime_1.jsx("line", { x1: endPoint.x, y1: endPoint.y, x2: secondHandle.x, y2: secondHandle.y, stroke: "rgba(255, 255, 255, 0.35)", strokeWidth: 1 }), jsx_runtime_1.jsx("path", { d: path, fill: "none", stroke: colors_1.BLUE, strokeWidth: 3 }), jsx_runtime_1.jsx("circle", { cx: startPoint.x, cy: startPoint.y, r: 4, fill: "white" }), jsx_runtime_1.jsx("circle", { cx: endPoint.x, cy: endPoint.y, r: 4, fill: "white" }), jsx_runtime_1.jsx("circle", { cx: firstHandle.x, cy: firstHandle.y, r: 6, fill: "white", stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke", pointerEvents: "all", cursor: activeHandle === 0 ? 'grabbing' : 'default', onPointerDown: (event) => onHandlePointerDown(0, event) }), jsx_runtime_1.jsx("circle", { cx: secondHandle.x, cy: secondHandle.y, r: 6, fill: "white", stroke: colors_1.BLUE, strokeWidth: 2, vectorEffect: "non-scaling-stroke", pointerEvents: "all", cursor: activeHandle === 1 ? 'grabbing' : 'default', onPointerDown: (event) => onHandlePointerDown(1, event) }), jsx_runtime_1.jsx("text", { x: PLOT_LEFT - 22, y: yZero + 3, fill: colors_1.LIGHT_TEXT, fontSize: 9, children: "0" }), jsx_runtime_1.jsx("text", { x: PLOT_LEFT - 22, y: yOne + 3, fill: colors_1.LIGHT_TEXT, fontSize: 9, children: "1" })
                                ] }), jsx_runtime_1.jsxs("div", { style: coordinatesGrid, children: [
                                    jsx_runtime_1.jsxs("div", { style: coordinateRow, children: [
                                            jsx_runtime_1.jsx("div", { style: coordinateLabel, children: "X1" }), jsx_runtime_1.jsx("div", { style: coordinateInputWrapper, children: jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: bezier[0], status: "ok", onValueChange: (value) => setCoordinate(0, 'x', value), onValueChangeEnd: (value) => setCoordinate(0, 'x', value), onTextChange: () => undefined, min: 0, max: 1, step: 0.01, formatter: formatNumber, rightAlign: false, style: numberInputStyle, snapToStep: false }) })
                                        ] }), jsx_runtime_1.jsxs("div", { style: coordinateRow, children: [
                                            jsx_runtime_1.jsx("div", { style: coordinateLabel, children: "Y1" }), jsx_runtime_1.jsx("div", { style: coordinateInputWrapper, children: jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: bezier[1], status: "ok", onValueChange: (value) => setCoordinate(0, 'y', value), onValueChangeEnd: (value) => setCoordinate(0, 'y', value), onTextChange: () => undefined, min: Y_MIN, max: Y_MAX, step: 0.01, formatter: formatNumber, rightAlign: false, style: numberInputStyle, snapToStep: false }) })
                                        ] }), jsx_runtime_1.jsxs("div", { style: coordinateRow, children: [
                                            jsx_runtime_1.jsx("div", { style: coordinateLabel, children: "X2" }), jsx_runtime_1.jsx("div", { style: coordinateInputWrapper, children: jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: bezier[2], status: "ok", onValueChange: (value) => setCoordinate(1, 'x', value), onValueChangeEnd: (value) => setCoordinate(1, 'x', value), onTextChange: () => undefined, min: 0, max: 1, step: 0.01, formatter: formatNumber, rightAlign: false, style: numberInputStyle, snapToStep: false }) })
                                        ] }), jsx_runtime_1.jsxs("div", { style: coordinateRow, children: [
                                            jsx_runtime_1.jsx("div", { style: coordinateLabel, children: "Y2" }), jsx_runtime_1.jsx("div", { style: coordinateInputWrapper, children: jsx_runtime_1.jsx(InputDragger_1.InputDragger, { type: "number", value: bezier[3], status: "ok", onValueChange: (value) => setCoordinate(1, 'y', value), onValueChangeEnd: (value) => setCoordinate(1, 'y', value), onTextChange: () => undefined, min: Y_MIN, max: Y_MAX, step: 0.01, formatter: formatNumber, rightAlign: false, style: numberInputStyle, snapToStep: false }) })
                                        ] })
                                ] })
                        ] }), jsx_runtime_1.jsx(ModalFooter_1.ModalFooterContainer, { children: jsx_runtime_1.jsxs(layout_1.Row, { justify: "flex-end", align: "center", children: [
                                jsx_runtime_1.jsx(Button_1.Button, { onClick: close, children: "Discard" }), jsx_runtime_1.jsx(layout_1.Spacing, { x: 1 }), jsx_runtime_1.jsx(ModalButton_1.ModalButton, { onClick: onSave, disabled: saveDisabled, children: "Save" })
                            ] }) }), jsx_runtime_1.jsx("button", { type: "submit", style: hiddenSubmit, disabled: saveDisabled, "aria-hidden": true, tabIndex: -1 })
                ] })
        ] }));
};
exports.EasingEditorModal = EasingEditorModal;
