// @remotion/studio/dist/components/Timeline/TimelineSelection.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useTimelineRowContainsSelection = exports.useTimelineGuideSelection = exports.useTimelineEasingSelection = exports.useTimelineKeyframeSelection = exports.useTimelineRowSelection = exports.useTimelineMarqueeSelectableItem = exports.useTimelineMarqueeSelection = exports.useCurrentTimelineSelectionStateAsRef = exports.TIMELINE_SCRUBBER_ATTR = exports.TIMELINE_MARQUEE_ITEM_ATTR = exports.useTimelineSelection = exports.TimelineSelectionProvider = exports.TimelineSelectAllKeybindings = exports.getTimelineSequenceSelectionKey = exports.getSelectableTimelineSequenceSelections = exports.getTimelineSelectionKey = exports.getTimelineSelectionFromNodePathInfo = exports.getTimelineMarqueeSelection = exports.timelineMarqueeRectsIntersect = exports.getClampedTimelineMarqueePoint = exports.getNormalizedTimelineMarqueeRect = exports.getTimelineSelectionAfterInteraction = exports.shouldSelectTimelineRowOnPointerDown = exports.isTimelineSelectionModifierEvent = exports.TIMELINE_TICKS_BACKGROUND = exports.TIMELINE_BACKGROUND = exports.getTimelineSelectedTrackHighlightStyle = exports.getTimelineColor = exports.getTimelineSelectedLabelStyle = exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING = exports.TIMELINE_SELECTED_LABEL_TEXT = exports.TIMELINE_SELECTED_LABEL_BACKGROUND = exports.TIMELINE_SELECTED_BACKGROUND = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const studio_shared_1 = require("@remotion/studio-shared");
const react_1 = require("react");
const client_id_1 = require("../../helpers/client-id");
const colors_1 = require("../../helpers/colors");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
const use_keybinding_1 = require("../../helpers/use-keybinding");
const z_index_1 = require("../../state/z-index");
const TimelineClipboardKeybindings_1 = require("./TimelineClipboardKeybindings");
const TimelineDeleteKeybindings_1 = require("./TimelineDeleteKeybindings");
exports.TIMELINE_SELECTED_BACKGROUND = '#3B3F42';
exports.TIMELINE_SELECTED_LABEL_BACKGROUND = '#B0B0B0';
exports.TIMELINE_SELECTED_LABEL_TEXT = 'black';
exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING = 2;
const getTimelineSelectedLabelStyle = (selected, subcategory) => {
    return {
        paddingLeft: exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING,
        paddingRight: exports.TIMELINE_SELECTED_LABEL_HORIZONTAL_PADDING,
        ...(selected
            ? {
                backgroundColor: subcategory
                    ? 'rgba(255, 255, 255, 0.1)'
                    : exports.TIMELINE_SELECTED_LABEL_BACKGROUND,
            }
            : {}),
    };
};
exports.getTimelineSelectedLabelStyle = getTimelineSelectedLabelStyle;
const getTimelineColor = (selected, subcategory) => {
    return selected && !subcategory
        ? exports.TIMELINE_SELECTED_LABEL_TEXT
        : 'rgba(255, 255, 255, 0.8)';
};
exports.getTimelineColor = getTimelineColor;
const getTimelineSelectedTrackHighlightStyle = (timelineWidth) => ({
    backgroundColor: exports.TIMELINE_SELECTED_BACKGROUND,
    bottom: 0,
    left: -timeline_layout_1.TIMELINE_PADDING,
    pointerEvents: 'none',
    position: 'absolute',
    top: 0,
    width: timelineWidth,
});
exports.getTimelineSelectedTrackHighlightStyle = getTimelineSelectedTrackHighlightStyle;
exports.TIMELINE_BACKGROUND = '#0F1113';
exports.TIMELINE_TICKS_BACKGROUND = colors_1.BACKGROUND;
const isTimelineSelectionModifierEvent = ({ shiftKey, metaKey, ctrlKey, }) => {
    return shiftKey || metaKey || ctrlKey;
};
exports.isTimelineSelectionModifierEvent = isTimelineSelectionModifierEvent;
const shouldSelectTimelineRowOnPointerDown = ({ selected, shiftKey, metaKey, ctrlKey, }) => {
    return (!selected || (0, exports.isTimelineSelectionModifierEvent)({ shiftKey, metaKey, ctrlKey }));
};
exports.shouldSelectTimelineRowOnPointerDown = shouldSelectTimelineRowOnPointerDown;
const getTimelineSelectionType = (item) => item.type;
const areTimelineSelectionTypesCompatible = (firstType, secondType) => {
    if (firstType === secondType) {
        return true;
    }
    return ((firstType === 'sequence-prop' && secondType === 'sequence-effect-prop') ||
        (firstType === 'sequence-effect-prop' && secondType === 'sequence-prop') ||
        (firstType === 'keyframe' && secondType === 'easing') ||
        (firstType === 'easing' && secondType === 'keyframe'));
};
const isTimelineSelectionCompatibleWithType = (item, type) => areTimelineSelectionTypesCompatible(getTimelineSelectionType(item), type);
const getTimelineSelectionAnchor = (selectedItems, previousAnchor, targetType) => {
    if (previousAnchor &&
        getTimelineSelectionType(previousAnchor) === targetType) {
        return previousAnchor;
    }
    for (let i = selectedItems.length - 1; i >= 0; i--) {
        const candidate = selectedItems[i];
        if (getTimelineSelectionType(candidate) === targetType) {
            return candidate;
        }
    }
    return null;
};
const getRangeSelection = ({ anchor, clickedItem, allSelectableItems, }) => {
    const anchorKey = (0, exports.getTimelineSelectionKey)(anchor);
    const clickedKey = (0, exports.getTimelineSelectionKey)(clickedItem);
    const orderedOfType = allSelectableItems.filter((item) => getTimelineSelectionType(item) === clickedItem.type);
    const anchorIndex = orderedOfType.findIndex((item) => (0, exports.getTimelineSelectionKey)(item) === anchorKey);
    const clickedIndex = orderedOfType.findIndex((item) => (0, exports.getTimelineSelectionKey)(item) === clickedKey);
    if (anchorIndex === -1 || clickedIndex === -1) {
        return [clickedItem];
    }
    const [from, to] = anchorIndex < clickedIndex
        ? [anchorIndex, clickedIndex]
        : [clickedIndex, anchorIndex];
    return orderedOfType.slice(from, to + 1);
};
const getTimelineSelectionAfterInteraction = ({ currentState, clickedItem, interaction, allSelectableItems, }) => {
    const { selectedItems, anchor: previousAnchor } = currentState;
    const clickedType = getTimelineSelectionType(clickedItem);
    if (clickedType === 'guide') {
        return {
            selectedItems: [clickedItem],
            anchor: clickedItem,
        };
    }
    const nextAnchor = getTimelineSelectionAnchor(selectedItems, previousAnchor, clickedType);
    const clickedKey = (0, exports.getTimelineSelectionKey)(clickedItem);
    if (interaction.shiftKey && nextAnchor) {
        return {
            selectedItems: getRangeSelection({
                anchor: nextAnchor,
                clickedItem,
                allSelectableItems,
            }),
            anchor: nextAnchor,
        };
    }
    if (interaction.toggleKey) {
        const compatibleItems = selectedItems.filter((item) => isTimelineSelectionCompatibleWithType(item, clickedType));
        const existingKeySet = new Set(compatibleItems.map(exports.getTimelineSelectionKey));
        if (existingKeySet.has(clickedKey)) {
            const toggledSelection = compatibleItems.filter((item) => (0, exports.getTimelineSelectionKey)(item) !== clickedKey);
            return {
                selectedItems: toggledSelection,
                anchor: toggledSelection.length === 0 ? null : clickedItem,
            };
        }
        const selectableOrderMap = new Map(allSelectableItems
            .filter((item) => isTimelineSelectionCompatibleWithType(item, clickedType))
            .map((item, index) => [(0, exports.getTimelineSelectionKey)(item), index]));
        const extendedSelection = [...compatibleItems, clickedItem].sort((a, b) => {
            var _a, _b;
            return (((_a = selectableOrderMap.get((0, exports.getTimelineSelectionKey)(a))) !== null && _a !== void 0 ? _a : 0) -
                ((_b = selectableOrderMap.get((0, exports.getTimelineSelectionKey)(b))) !== null && _b !== void 0 ? _b : 0));
        });
        return {
            selectedItems: extendedSelection,
            anchor: clickedItem,
        };
    }
    return {
        selectedItems: [clickedItem],
        anchor: clickedItem,
    };
};
exports.getTimelineSelectionAfterInteraction = getTimelineSelectionAfterInteraction;
const getNormalizedTimelineMarqueeRect = ({ startX, startY, currentX, currentY, }) => ({
    left: Math.min(startX, currentX),
    top: Math.min(startY, currentY),
    right: Math.max(startX, currentX),
    bottom: Math.max(startY, currentY),
});
exports.getNormalizedTimelineMarqueeRect = getNormalizedTimelineMarqueeRect;
const getClampedTimelineMarqueePoint = ({ x, y, bounds, }) => ({
    x: Math.min(bounds.right, Math.max(bounds.left, x)),
    y: Math.min(bounds.bottom, Math.max(bounds.top, y)),
});
exports.getClampedTimelineMarqueePoint = getClampedTimelineMarqueePoint;
const timelineMarqueeRectsIntersect = (a, b) => a.left <= b.right &&
    a.right >= b.left &&
    a.top <= b.bottom &&
    a.bottom >= b.top;
exports.timelineMarqueeRectsIntersect = timelineMarqueeRectsIntersect;
const getTimelineMarqueeSelectionKind = (item) => {
    if (item.type === 'sequence') {
        return 'sequence';
    }
    if (item.type === 'keyframe' || item.type === 'easing') {
        return 'keyframes-and-easings';
    }
    return null;
};
const isTimelineSelectionCompatibleWithMarqueeKind = (item, kind) => {
    if (kind === 'sequence') {
        return item.type === 'sequence';
    }
    return item.type === 'keyframe' || item.type === 'easing';
};
const getTimelineMarqueeSelection = ({ candidates, lockedSelectionKind, marqueeRect, }) => {
    const intersectingCandidates = candidates.filter((candidate) => {
        return (getTimelineMarqueeSelectionKind(candidate.item) !== null &&
            (0, exports.timelineMarqueeRectsIntersect)(candidate.rect, marqueeRect));
    });
    const getFirstIntersectingSelectionKind = () => intersectingCandidates.length === 0
        ? null
        : getTimelineMarqueeSelectionKind(intersectingCandidates[0].item);
    let nextLockedSelectionKind = lockedSelectionKind !== null && lockedSelectionKind !== void 0 ? lockedSelectionKind : getFirstIntersectingSelectionKind();
    const getSelectedItemsForKind = (kind) => intersectingCandidates
        .filter((candidate) => isTimelineSelectionCompatibleWithMarqueeKind(candidate.item, kind))
        .map((candidate) => candidate.item);
    if (nextLockedSelectionKind === null) {
        return { lockedSelectionKind: null, selectedItems: [] };
    }
    let selectedItems = getSelectedItemsForKind(nextLockedSelectionKind);
    if (lockedSelectionKind !== null && selectedItems.length === 0) {
        nextLockedSelectionKind = getFirstIntersectingSelectionKind();
        selectedItems =
            nextLockedSelectionKind === null
                ? []
                : getSelectedItemsForKind(nextLockedSelectionKind);
    }
    return {
        lockedSelectionKind: nextLockedSelectionKind,
        selectedItems,
    };
};
exports.getTimelineMarqueeSelection = getTimelineMarqueeSelection;
const defaultTimelineSelectionContextValue = {
    canSelect: false,
    selectedItems: [],
    isSelected: () => false,
    selectItem: () => undefined,
    selectItems: () => undefined,
    registerSelectableItem: () => () => undefined,
    registerMarqueeSelectableItem: () => () => undefined,
    getMarqueeSelection: () => ({
        lockedSelectionKind: null,
        selectedItems: [],
    }),
    containsSelection: () => false,
    clearSelection: () => undefined,
};
const TimelineSelectionContext = (0, react_1.createContext)(defaultTimelineSelectionContextValue);
const CurrentTimelineSelectionContext = (0, react_1.createContext)(null);
const parseEffectIndex = (effectIndex) => {
    const parsed = Number(effectIndex);
    if (!Number.isInteger(parsed) || parsed < 0) {
        return null;
    }
    return parsed;
};
const getTimelineSelectionFromNodePathInfo = (nodePathInfo) => {
    if (nodePathInfo === null) {
        return null;
    }
    const { auxiliaryKeys } = nodePathInfo;
    if (auxiliaryKeys.length === 0) {
        return { type: 'sequence', nodePathInfo };
    }
    if (auxiliaryKeys.length === 2 && auxiliaryKeys[0] === 'controls') {
        return { type: 'sequence-prop', nodePathInfo, key: auxiliaryKeys[1] };
    }
    if (auxiliaryKeys.length === 1 && auxiliaryKeys[0] === 'effects') {
        return { type: 'sequence-all-effects', nodePathInfo };
    }
    if (auxiliaryKeys[0] === 'effects') {
        const effectIndex = parseEffectIndex(auxiliaryKeys[1]);
        if (effectIndex === null) {
            return null;
        }
        if (auxiliaryKeys.length === 2) {
            return { type: 'sequence-effect', nodePathInfo, i: effectIndex };
        }
        if (auxiliaryKeys.length === 3) {
            return {
                type: 'sequence-effect-prop',
                nodePathInfo,
                i: effectIndex,
                key: auxiliaryKeys[2],
            };
        }
    }
    return null;
};
exports.getTimelineSelectionFromNodePathInfo = getTimelineSelectionFromNodePathInfo;
const getTimelineSelectionKey = (item) => {
    switch (item.type) {
        case 'guide':
            return `guide.${item.guideId}`;
        case 'sequence':
            return `${(0, exports.getTimelineSequenceSelectionKey)(item.nodePathInfo)}.sequence`;
        case 'sequence-prop':
            return `${(0, exports.getTimelineSequenceSelectionKey)(item.nodePathInfo)}.sequence-prop.${item.key}`;
        case 'sequence-all-effects':
            return `${(0, exports.getTimelineSequenceSelectionKey)(item.nodePathInfo)}.sequence-all-effects`;
        case 'sequence-effect':
            return `${(0, exports.getTimelineSequenceSelectionKey)(item.nodePathInfo)}.sequence-effect.${item.i}`;
        case 'sequence-effect-prop':
            return `${(0, exports.getTimelineSequenceSelectionKey)(item.nodePathInfo)}.sequence-effect-prop.${item.i}.${item.key}`;
        case 'keyframe':
            return `${(0, timeline_node_path_key_1.timelineNodePathInfoToKey)(item.nodePathInfo)}.keyframe.${item.frame}`;
        case 'easing':
            return `${(0, timeline_node_path_key_1.timelineNodePathInfoToKey)(item.nodePathInfo)}.easing.${item.segmentIndex}`;
        default:
            throw new Error(`Unexpected timeline selection type: ${item}`);
    }
};
exports.getTimelineSelectionKey = getTimelineSelectionKey;
const nodePathDescendsFrom = (descendant, ancestor) => {
    if ((0, studio_shared_1.stringifySequenceExpandedRowKey)(descendant.sequenceSubscriptionKey) !==
        (0, studio_shared_1.stringifySequenceExpandedRowKey)(ancestor.sequenceSubscriptionKey)) {
        return false;
    }
    if (descendant.index !== ancestor.index) {
        return false;
    }
    // Must be strictly deeper than `ancestor` (i.e. a descendant), not the same row.
    if (descendant.auxiliaryKeys.length <= ancestor.auxiliaryKeys.length) {
        return false;
    }
    return ancestor.auxiliaryKeys.every((key, i) => descendant.auxiliaryKeys[i] === key);
};
const getSelectableTimelineSequenceSelections = (tracks) => {
    return tracks.flatMap((track) => {
        if (track.nodePathInfo === null ||
            track.nodePathInfo.auxiliaryKeys.length > 0) {
            return [];
        }
        return [{ type: 'sequence', nodePathInfo: track.nodePathInfo }];
    });
};
exports.getSelectableTimelineSequenceSelections = getSelectableTimelineSequenceSelections;
const getTimelineSequenceSelectionKey = (nodePathInfo) => (0, timeline_node_path_key_1.timelineNodePathInfoToKey)({ ...nodePathInfo, auxiliaryKeys: [] });
exports.getTimelineSequenceSelectionKey = getTimelineSequenceSelectionKey;
const TimelineSelectAllKeybindings = ({ timeline }) => {
    const keybindings = (0, use_keybinding_1.useKeybinding)();
    const { canSelect } = (0, exports.useTimelineSelection)();
    const currentSelection = (0, exports.useCurrentTimelineSelectionStateAsRef)();
    const selectableSequenceSelections = (0, react_1.useMemo)(() => (0, exports.getSelectableTimelineSequenceSelections)(timeline), [timeline]);
    const selectableSequenceSelectionsRef = (0, react_1.useRef)(selectableSequenceSelections);
    selectableSequenceSelectionsRef.current = selectableSequenceSelections;
    (0, react_1.useEffect)(() => {
        if (!canSelect) {
            return;
        }
        const selectAll = keybindings.registerKeybinding({
            event: 'keydown',
            key: 'a',
            callback: () => {
                const latestSelectableSequenceSelections = selectableSequenceSelectionsRef.current;
                if (latestSelectableSequenceSelections.length === 0) {
                    return;
                }
                currentSelection.current.selectItems(latestSelectableSequenceSelections);
            },
            commandCtrlKey: true,
            preventDefault: true,
            triggerIfInputFieldFocused: false,
            keepRegisteredWhenNotHighestContext: false,
        });
        return () => {
            selectAll.unregister();
        };
    }, [canSelect, currentSelection, keybindings]);
    return null;
};
exports.TimelineSelectAllKeybindings = TimelineSelectAllKeybindings;
const TimelineSelectionProvider = ({ children }) => {
    const { previewServerState } = (0, react_1.useContext)(client_id_1.StudioServerConnectionCtx);
    const canSelect = previewServerState.type === 'connected' &&
        !window.remotion_isReadOnlyStudio;
    const [selectedItems, setSelectedItems] = (0, react_1.useState)([]);
    const selectionAnchor = (0, react_1.useRef)(null);
    const selectableItemsOrder = (0, react_1.useRef)(new Map());
    const selectableItems = (0, react_1.useRef)(new Map());
    const marqueeSelectableItems = (0, react_1.useRef)(new Map());
    const registrationCounter = (0, react_1.useRef)(0);
    const marqueeRegistrationCounter = (0, react_1.useRef)(0);
    (0, react_1.useEffect)(() => {
        if (!canSelect) {
            setSelectedItems([]);
        }
    }, [canSelect]);
    const canSelectItem = (0, react_1.useCallback)((_item) => canSelect, [canSelect]);
    const selectedKeys = (0, react_1.useMemo)(() => new Set(selectedItems.map(exports.getTimelineSelectionKey)), [selectedItems]);
    const isSelected = (0, react_1.useCallback)((item) => {
        return selectedKeys.has((0, exports.getTimelineSelectionKey)(item));
    }, [selectedKeys]);
    const selectItem = (0, react_1.useCallback)((item, interaction = {
        shiftKey: false,
        toggleKey: false,
    }) => {
        if (!canSelectItem(item)) {
            return;
        }
        setSelectedItems((currentSelectedItems) => {
            const orderedSelectableItems = [
                ...selectableItems.current.values(),
            ].sort((a, b) => {
                var _a, _b;
                return (((_a = selectableItemsOrder.current.get((0, exports.getTimelineSelectionKey)(a))) !== null && _a !== void 0 ? _a : 0) -
                    ((_b = selectableItemsOrder.current.get((0, exports.getTimelineSelectionKey)(b))) !== null && _b !== void 0 ? _b : 0));
            });
            const nextState = (0, exports.getTimelineSelectionAfterInteraction)({
                currentState: {
                    selectedItems: currentSelectedItems,
                    anchor: selectionAnchor.current,
                },
                clickedItem: item,
                interaction,
                allSelectableItems: orderedSelectableItems,
            });
            selectionAnchor.current = nextState.anchor;
            return nextState.selectedItems;
        });
    }, [canSelectItem]);
    const selectItems = (0, react_1.useCallback)((items) => {
        if (!items.every(canSelectItem)) {
            return;
        }
        selectionAnchor.current =
            items.length === 0 ? null : items[items.length - 1];
        setSelectedItems(items);
    }, [canSelectItem]);
    const registerSelectableItem = (0, react_1.useCallback)((item) => {
        const key = (0, exports.getTimelineSelectionKey)(item);
        const registrationOrder = registrationCounter.current;
        registrationCounter.current += 1;
        selectableItems.current.set(key, item);
        selectableItemsOrder.current.set(key, registrationOrder);
        return () => {
            selectableItems.current.delete(key);
            selectableItemsOrder.current.delete(key);
        };
    }, []);
    const registerMarqueeSelectableItem = (0, react_1.useCallback)((item, getRect) => {
        const key = (0, exports.getTimelineSelectionKey)(item);
        const registrationOrder = marqueeRegistrationCounter.current;
        marqueeRegistrationCounter.current += 1;
        marqueeSelectableItems.current.set(key, {
            getRect,
            item,
            order: registrationOrder,
        });
        return () => {
            marqueeSelectableItems.current.delete(key);
        };
    }, []);
    const getMarqueeSelectionForRect = (0, react_1.useCallback)((marqueeRect, lockedSelectionKind) => {
        const candidates = [...marqueeSelectableItems.current.values()]
            .sort((a, b) => a.order - b.order)
            .flatMap((candidate) => {
            if (!canSelectItem(candidate.item)) {
                return [];
            }
            const rect = candidate.getRect();
            if (rect === null) {
                return [];
            }
            return [
                {
                    item: candidate.item,
                    rect: {
                        bottom: rect.bottom,
                        left: rect.left,
                        right: rect.right,
                        top: rect.top,
                    },
                },
            ];
        });
        return (0, exports.getTimelineMarqueeSelection)({
            candidates,
            lockedSelectionKind,
            marqueeRect,
        });
    }, [canSelectItem]);
    const clearSelection = (0, react_1.useCallback)(() => {
        selectionAnchor.current = null;
        setSelectedItems([]);
    }, []);
    const containsSelection = (0, react_1.useCallback)((nodePathInfo) => {
        return selectedItems.some((selected) => selected.type !== 'guide' &&
            nodePathDescendsFrom(selected.nodePathInfo, nodePathInfo));
    }, [selectedItems]);
    const value = (0, react_1.useMemo)(() => ({
        canSelect,
        selectedItems,
        isSelected,
        selectItem,
        selectItems,
        registerSelectableItem,
        registerMarqueeSelectableItem,
        getMarqueeSelection: getMarqueeSelectionForRect,
        containsSelection,
        clearSelection,
    }), [
        canSelect,
        selectedItems,
        isSelected,
        selectItem,
        selectItems,
        registerSelectableItem,
        registerMarqueeSelectableItem,
        getMarqueeSelectionForRect,
        containsSelection,
        clearSelection,
    ]);
    const currentSelection = (0, react_1.useRef)(value);
    currentSelection.current = value;
    return (jsx_runtime_1.jsx(CurrentTimelineSelectionContext.Provider, { value: currentSelection, children: jsx_runtime_1.jsxs(TimelineSelectionContext.Provider, { value: value, children: [children, jsx_runtime_1.jsx(TimelineClipboardKeybindings_1.TimelineClipboardKeybindings, {}), jsx_runtime_1.jsx(TimelineDeleteKeybindings_1.TimelineDeleteKeybindings, {})
            ] }) }));
};
exports.TimelineSelectionProvider = TimelineSelectionProvider;
const useTimelineSelection = () => {
    return (0, react_1.useContext)(TimelineSelectionContext);
};
exports.useTimelineSelection = useTimelineSelection;
exports.TIMELINE_MARQUEE_ITEM_ATTR = 'data-timeline-marquee-item';
exports.TIMELINE_SCRUBBER_ATTR = 'data-timeline-scrubber';
const useCurrentTimelineSelectionStateAsRef = () => {
    const currentSelection = (0, react_1.useContext)(CurrentTimelineSelectionContext);
    if (currentSelection === null) {
        throw new Error('useCurrentTimelineSelectionStateAsRef must be used inside TimelineSelectionProvider');
    }
    return currentSelection;
};
exports.useCurrentTimelineSelectionStateAsRef = useCurrentTimelineSelectionStateAsRef;
const useTimelineMarqueeSelection = () => {
    const { canSelect, getMarqueeSelection, selectItems } = (0, exports.useTimelineSelection)();
    const { isHighestContext } = (0, z_index_1.useZIndex)();
    const [marqueeRect, setMarqueeRect] = (0, react_1.useState)(null);
    const onPointerDownCapture = (0, react_1.useCallback)((event) => {
        if (!isHighestContext) {
            return;
        }
        if (event.button !== 0 || !canSelect) {
            return;
        }
        if (event.shiftKey || event.metaKey || event.ctrlKey) {
            return;
        }
        if (!(event.target instanceof Element)) {
            return;
        }
        if (event.target.closest(`[${exports.TIMELINE_MARQUEE_ITEM_ATTR}]`) ||
            event.target.closest(`[${exports.TIMELINE_SCRUBBER_ATTR}]`)) {
            return;
        }
        const { currentTarget: target, pointerId } = event;
        if (target.setPointerCapture) {
            target.setPointerCapture(pointerId);
        }
        const initialBounds = target.getBoundingClientRect();
        const marqueeBounds = {
            bottom: initialBounds.bottom,
            left: initialBounds.left,
            right: initialBounds.right,
            top: initialBounds.top,
        };
        const start = (0, exports.getClampedTimelineMarqueePoint)({
            bounds: marqueeBounds,
            x: event.clientX,
            y: event.clientY,
        });
        const startX = start.x;
        const startY = start.y;
        const previousUserSelect = document.body.style.userSelect;
        const previousWebkitUserSelect = document.body.style.webkitUserSelect;
        document.body.style.userSelect = 'none';
        document.body.style.webkitUserSelect = 'none';
        let hasDragged = false;
        let lockedSelectionKind = null;
        const cleanup = () => {
            var _a;
            window.removeEventListener('pointermove', onPointerMove);
            window.removeEventListener('pointerup', onPointerUp);
            window.removeEventListener('pointercancel', onPointerCancel);
            if ((_a = target.hasPointerCapture) === null || _a === void 0 ? void 0 : _a.call(target, pointerId)) {
                target.releasePointerCapture(pointerId);
            }
            document.body.style.userSelect = previousUserSelect;
            document.body.style.webkitUserSelect = previousWebkitUserSelect;
            setMarqueeRect(null);
        };
        const updateSelection = (clientX, clientY) => {
            const current = (0, exports.getClampedTimelineMarqueePoint)({
                bounds: marqueeBounds,
                x: clientX,
                y: clientY,
            });
            if (!hasDragged &&
                Math.max(Math.abs(current.x - startX), Math.abs(current.y - startY)) <
                    3) {
                return;
            }
            hasDragged = true;
            const rect = (0, exports.getNormalizedTimelineMarqueeRect)({
                currentX: current.x,
                currentY: current.y,
                startX,
                startY,
            });
            const nextSelection = getMarqueeSelection(rect, lockedSelectionKind);
            lockedSelectionKind = nextSelection.lockedSelectionKind;
            setMarqueeRect(rect);
            selectItems(nextSelection.selectedItems);
        };
        const onPointerMove = (moveEvent) => {
            updateSelection(moveEvent.clientX, moveEvent.clientY);
        };
        const onPointerUp = (upEvent) => {
            updateSelection(upEvent.clientX, upEvent.clientY);
            cleanup();
        };
        const onPointerCancel = () => {
            cleanup();
        };
        window.addEventListener('pointermove', onPointerMove);
        window.addEventListener('pointerup', onPointerUp);
        window.addEventListener('pointercancel', onPointerCancel);
    }, [canSelect, getMarqueeSelection, isHighestContext, selectItems]);
    return { marqueeRect, onPointerDownCapture };
};
exports.useTimelineMarqueeSelection = useTimelineMarqueeSelection;
const useTimelineMarqueeSelectableItem = (item, ref) => {
    const { registerMarqueeSelectableItem } = (0, exports.useTimelineSelection)();
    (0, react_1.useEffect)(() => {
        if (item === null) {
            return;
        }
        return registerMarqueeSelectableItem(item, () => {
            var _a;
            var _b;
            return (_b = (_a = ref.current) === null || _a === void 0 ? void 0 : _a.getBoundingClientRect()) !== null && _b !== void 0 ? _b : null;
        });
    }, [item, ref, registerMarqueeSelectableItem]);
};
exports.useTimelineMarqueeSelectableItem = useTimelineMarqueeSelectableItem;
const useTimelineRowSelection = (nodePathInfo) => {
    const { canSelect, isSelected, selectItem, registerSelectableItem } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => (0, exports.getTimelineSelectionFromNodePathInfo)(nodePathInfo), [nodePathInfo]);
    (0, react_1.useEffect)(() => {
        if (selectionItem === null) {
            return;
        }
        return registerSelectableItem(selectionItem);
    }, [registerSelectableItem, selectionItem]);
    const selected = selectionItem === null ? false : isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)((interaction) => {
        if (selectionItem === null) {
            return;
        }
        selectItem(selectionItem, interaction);
    }, [selectItem, selectionItem]);
    return {
        onSelect,
        selectable: canSelect && selectionItem !== null,
        selectionItem,
        selected,
    };
};
exports.useTimelineRowSelection = useTimelineRowSelection;
const useTimelineKeyframeSelection = (nodePathInfo, frame) => {
    const { canSelect, isSelected, selectItem, registerSelectableItem } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => ({
        type: 'keyframe',
        nodePathInfo,
        frame,
    }), [nodePathInfo, frame]);
    (0, react_1.useEffect)(() => {
        return registerSelectableItem(selectionItem);
    }, [registerSelectableItem, selectionItem]);
    const selected = isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)((interaction) => {
        selectItem(selectionItem, interaction);
    }, [selectItem, selectionItem]);
    return {
        onSelect,
        selectable: canSelect,
        selectionItem,
        selected,
    };
};
exports.useTimelineKeyframeSelection = useTimelineKeyframeSelection;
const useTimelineEasingSelection = ({ nodePathInfo, fromFrame, toFrame, segmentIndex, }) => {
    const { canSelect, isSelected, selectItem, registerSelectableItem } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => ({
        type: 'easing',
        nodePathInfo,
        fromFrame,
        toFrame,
        segmentIndex,
    }), [nodePathInfo, fromFrame, segmentIndex, toFrame]);
    (0, react_1.useEffect)(() => {
        return registerSelectableItem(selectionItem);
    }, [registerSelectableItem, selectionItem]);
    const selected = isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)((interaction) => {
        selectItem(selectionItem, interaction);
    }, [selectItem, selectionItem]);
    return {
        onSelect,
        selectable: canSelect,
        selected,
        selectionItem,
    };
};
exports.useTimelineEasingSelection = useTimelineEasingSelection;
const useTimelineGuideSelection = (guideId) => {
    const { canSelect, clearSelection, isSelected, selectItem, registerSelectableItem, } = (0, exports.useTimelineSelection)();
    const selectionItem = (0, react_1.useMemo)(() => ({
        type: 'guide',
        guideId,
    }), [guideId]);
    (0, react_1.useEffect)(() => {
        return registerSelectableItem(selectionItem);
    }, [registerSelectableItem, selectionItem]);
    const selected = isSelected(selectionItem);
    const onSelect = (0, react_1.useCallback)(() => {
        selectItem(selectionItem);
    }, [selectItem, selectionItem]);
    return {
        clearSelection,
        onSelect,
        selectable: canSelect,
        selected,
        selectionItem,
    };
};
exports.useTimelineGuideSelection = useTimelineGuideSelection;
const useTimelineRowContainsSelection = (nodePathInfo) => {
    const { containsSelection } = (0, exports.useTimelineSelection)();
    if (nodePathInfo === null) {
        return false;
    }
    return containsSelection(nodePathInfo);
};
exports.useTimelineRowContainsSelection = useTimelineRowContainsSelection;
