// @remotion/studio/dist/components/Timeline/TimelineItemStack.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineItemStack = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const colors_1 = require("../../helpers/colors");
const source_attribution_1 = require("./TimelineStack/source-attribution");
const TimelineItemStack = ({ originalLocation }) => {
    const style = (0, react_1.useMemo)(() => {
        return {
            fontSize: 12,
            color: colors_1.VERY_LIGHT_TEXT,
            display: 'flex',
            flexDirection: 'row',
            alignItems: 'center',
            whiteSpace: 'nowrap',
            textOverflow: 'ellipsis',
            overflow: 'hidden',
            flexShrink: 100000,
            userSelect: 'none',
            WebkitUserSelect: 'none',
        };
    }, []);
    if (!originalLocation) {
        return null;
    }
    return (jsx_runtime_1.jsx("div", { style: style, children: (0, source_attribution_1.getOriginalSourceAttribution)(originalLocation) }));
};
exports.TimelineItemStack = TimelineItemStack;
