// @remotion/studio/dist/components/Timeline/TimelineKeyframedValue.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TimelineKeyframedValue = void 0;
const jsx_runtime_1 = require("react/jsx-runtime");
const react_1 = require("react");
const remotion_1 = require("remotion");
const TimelineSchemaField_1 = require("./TimelineSchemaField");
const valuesEqual = (left, right) => {
    return JSON.stringify(left) === JSON.stringify(right);
};
const TimelineKeyframedValue = ({ field, propStatus, keyframeDisplayOffset, dragOverrideValue, onSave, onDragValueChange, onDragEnd, scaleLockNodePath, }) => {
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const jsxFrame = timelinePosition - keyframeDisplayOffset;
    const computedValue = (0, react_1.useMemo)(() => {
        const raw = remotion_1.Internals.interpolateKeyframedStatus({
            frame: jsxFrame,
            status: propStatus,
        });
        if (typeof raw === 'number') {
            return Math.round(raw * 100) / 100;
        }
        return raw;
    }, [jsxFrame, propStatus]);
    const fakeStatus = (0, react_1.useMemo)(() => ({
        status: 'static',
        codeValue: computedValue,
    }), [computedValue]);
    const effectiveValue = (0, react_1.useMemo)(() => {
        return remotion_1.Internals.getEffectiveVisualModeValue({
            propStatus: fakeStatus,
            dragOverrideValue,
            frame: jsxFrame,
            defaultValue: field.fieldSchema.default,
            shouldResortToDefaultValueIfUndefined: true,
        });
    }, [dragOverrideValue, fakeStatus, field.fieldSchema.default, jsxFrame]);
    const onSaveIfChanged = (0, react_1.useCallback)((value) => {
        const existingKeyframe = propStatus.keyframes.find((keyframe) => keyframe.frame === jsxFrame);
        if (valuesEqual(value, computedValue) ||
            (existingKeyframe && valuesEqual(value, existingKeyframe.value))) {
            return Promise.resolve();
        }
        return onSave(value, jsxFrame);
    }, [computedValue, jsxFrame, onSave, propStatus.keyframes]);
    if (computedValue === null) {
        return null;
    }
    return (jsx_runtime_1.jsx(TimelineSchemaField_1.TimelineFieldValue, { field: field, propStatus: fakeStatus, effectiveValue: effectiveValue, onSave: onSaveIfChanged, onDragValueChange: onDragValueChange, onDragEnd: onDragEnd, scaleLockNodePath: scaleLockNodePath }));
};
exports.TimelineKeyframedValue = TimelineKeyframedValue;
