// @remotion/studio/dist/components/Timeline/use-expanded-track-keyframe-rows.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useExpandedTrackKeyframeRows = void 0;
const react_1 = require("react");
const remotion_1 = require("remotion");
const timeline_layout_1 = require("../../helpers/timeline-layout");
const timeline_node_path_key_1 = require("../../helpers/timeline-node-path-key");
const ExpandedTracksProvider_1 = require("../ExpandedTracksProvider");
const get_node_keyframes_1 = require("./get-node-keyframes");
const canEditEasingForInterpolationFunction = (interpolationFunction) => interpolationFunction === 'interpolate' ||
    interpolationFunction === 'interpolateColors';
const getNodeCanEditEasing = ({ node, nodePath, propStatuses, }) => {
    var _a;
    if (node.kind !== 'field' || node.field === null) {
        return false;
    }
    if (node.field.kind === 'sequence-field') {
        const sequencePropStatus = (_a = remotion_1.Internals.getPropStatusesCtx(propStatuses, nodePath)) === null || _a === void 0 ? void 0 : _a[node.field.key];
        return ((sequencePropStatus === null || sequencePropStatus === void 0 ? void 0 : sequencePropStatus.status) === 'keyframed' &&
            canEditEasingForInterpolationFunction(sequencePropStatus.interpolationFunction));
    }
    const effectStatus = remotion_1.Internals.getEffectPropStatusesCtx({
        propStatuses,
        nodePath,
        effectIndex: node.field.effectIndex,
    });
    const effectPropStatus = effectStatus.type === 'can-update-effect'
        ? effectStatus.props[node.field.key]
        : null;
    return ((effectPropStatus === null || effectPropStatus === void 0 ? void 0 : effectPropStatus.status) === 'keyframed' &&
        canEditEasingForInterpolationFunction(effectPropStatus.interpolationFunction));
};
const useExpandedTrackKeyframeRows = ({ sequence, nodePathInfo, keyframeDisplayOffset, }) => {
    const { getIsExpanded } = (0, react_1.useContext)(ExpandedTracksProvider_1.ExpandedTracksGetterContext);
    const { propStatuses } = (0, react_1.useContext)(remotion_1.Internals.VisualModePropStatusesContext);
    const { getDragOverrides, getEffectDragOverrides } = (0, react_1.useContext)(remotion_1.Internals.VisualModeDragOverridesContext);
    const timelinePosition = remotion_1.Internals.Timeline.useTimelinePosition();
    const tree = (0, react_1.useMemo)(() => (0, timeline_layout_1.buildTimelineTree)({
        sequence,
        nodePathInfo,
        getDragOverrides,
        getEffectDragOverrides,
        propStatuses,
    }), [
        propStatuses,
        getDragOverrides,
        getEffectDragOverrides,
        nodePathInfo,
        sequence,
    ]);
    const flat = (0, react_1.useMemo)(() => (0, timeline_layout_1.flattenVisibleTreeNodes)({ nodes: tree, getIsExpanded }), [tree, getIsExpanded]);
    const expandedHeight = (0, react_1.useMemo)(() => (0, timeline_layout_1.getExpandedTrackHeight)({
        sequence,
        nodePathInfo,
        getIsExpanded,
        propStatuses,
    }), [propStatuses, getIsExpanded, nodePathInfo, sequence]);
    const rows = (0, react_1.useMemo)(() => flat.map(({ node }) => ({
        height: (0, timeline_layout_1.getTreeRowHeight)(node),
        keyframes: (0, get_node_keyframes_1.getNodeKeyframes)({
            node,
            nodePath: nodePathInfo.sequenceSubscriptionKey,
            propStatuses,
            keyframeDisplayOffset,
            getDragOverrides,
            getEffectDragOverrides,
            timelinePosition,
        }),
        canEditEasing: getNodeCanEditEasing({
            node,
            nodePath: nodePathInfo.sequenceSubscriptionKey,
            propStatuses,
        }),
        rowKey: (0, timeline_node_path_key_1.timelineNodePathInfoToKey)(node.nodePathInfo),
        nodePathInfo: node.nodePathInfo,
    })), [
        propStatuses,
        flat,
        getDragOverrides,
        getEffectDragOverrides,
        keyframeDisplayOffset,
        nodePathInfo.sequenceSubscriptionKey,
        timelinePosition,
    ]);
    return { rows, expandedHeight };
};
exports.useExpandedTrackKeyframeRows = useExpandedTrackKeyframeRows;
