// @remotion/studio-shared/dist/shape-drag-data.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseShapeDragData = exports.makeShapeDragData = exports.areShapeAttributes = exports.isShapeAttribute = exports.isShapeAttributeName = exports.isShapeName = exports.shapeNames = exports.SHAPE_DRAG_MIME_TYPE = void 0;
exports.SHAPE_DRAG_MIME_TYPE = 'application/vnd.remotion.shape+json';
exports.shapeNames = [
    'Arrow',
    'Circle',
    'Ellipse',
    'Heart',
    'Pie',
    'Polygon',
    'Rect',
    'Star',
    'Triangle',
];
const isRecord = (value) => {
    return typeof value === 'object' && value !== null && !Array.isArray(value);
};
const isShapeName = (value) => {
    return typeof value === 'string' && exports.shapeNames.includes(value);
};
exports.isShapeName = isShapeName;
const isShapeAttributeName = (value) => {
    return (typeof value === 'string' &&
        value !== 'style' &&
        /^[A-Za-z_$][A-Za-z0-9_$]*$/.test(value));
};
exports.isShapeAttributeName = isShapeAttributeName;
const isShapeAttribute = (value) => {
    if (!isRecord(value)) {
        return false;
    }
    if (!(0, exports.isShapeAttributeName)(value.name)) {
        return false;
    }
    return (typeof value.value === 'string' ||
        typeof value.value === 'boolean' ||
        (typeof value.value === 'number' && Number.isFinite(value.value)));
};
exports.isShapeAttribute = isShapeAttribute;
const areShapeAttributes = (value) => {
    if (!Array.isArray(value)) {
        return false;
    }
    const seen = new Set();
    for (const attribute of value) {
        if (!(0, exports.isShapeAttribute)(attribute) || seen.has(attribute.name)) {
            return false;
        }
        seen.add(attribute.name);
    }
    return true;
};
exports.areShapeAttributes = areShapeAttributes;
const makeShapeDragData = ({ attributes, shape, }) => {
    return {
        type: 'remotion-shape',
        version: 1,
        shape,
        attributes,
    };
};
exports.makeShapeDragData = makeShapeDragData;
const parseShapeDragData = (value) => {
    try {
        const parsed = JSON.parse(value);
        if (!isRecord(parsed)) {
            return null;
        }
        if (parsed.type !== 'remotion-shape' || parsed.version !== 1) {
            return null;
        }
        if (!(0, exports.isShapeName)(parsed.shape)) {
            return null;
        }
        if (!(0, exports.areShapeAttributes)(parsed.attributes)) {
            return null;
        }
        return {
            type: 'remotion-shape',
            version: 1,
            shape: parsed.shape,
            attributes: parsed.attributes,
        };
    }
    catch (_a) {
        return null;
    }
};
exports.parseShapeDragData = parseShapeDragData;
