// @remotion/canvas-capture/dist/esm/index.mjs
var __create = Object.create;
var __getProtoOf = Object.getPrototypeOf;
var __defProp = Object.defineProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __toESM = (mod, isNodeMode, target) => {
  target = mod != null ? __create(__getProtoOf(mod)) : {};
  const to = isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target;
  for (let key of __getOwnPropNames(mod))
    if (!__hasOwnProp.call(to, key))
      __defProp(to, key, {
        get: () => mod[key],
        enumerable: true
      });
  return to;
};
var __require = /* @__PURE__ */ ((x) => typeof require !== "undefined" ? require : typeof Proxy !== "undefined" ? new Proxy(x, {
  get: (a, b) => (typeof require !== "undefined" ? require : a)[b]
}) : x)(function(x) {
  if (typeof require !== "undefined")
    return require.apply(this, arguments);
  throw Error('Dynamic require of "' + x + '" is not supported');
});

// src/index.tsx
import {
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useMemo,
  useRef
} from "react";
import { jsx } from "react/jsx-runtime";
var canvasStyle = {
  position: "absolute",
  inset: 0,
  width: "100%",
  height: "100%",
  display: "block"
};
var contentStyle = {
  position: "absolute",
  inset: 0,
  width: "100%",
  height: "100%",
  transformOrigin: "top left"
};
var fallbackFrameDurationInSeconds = 1 / 60;
var recordingVideoBitrate = 120000000;
var recordingKeyFrameIntervalInSeconds = 0.5;
var isHtmlInCanvasAvailable = () => {
  if (typeof document === "undefined") {
    return false;
  }
  const canvas = document.createElement("canvas");
  const context = canvas.getContext("2d");
  return typeof canvas.requestPaint === "function" && typeof context?.drawElementImage === "function";
};
var resetCanvas = (context, canvas) => {
  if (typeof context.reset === "function") {
    context.reset();
    return;
  }
  context.setTransform(1, 0, 0, 1, 0, 0);
  context.clearRect(0, 0, canvas.width, canvas.height);
};
var roundUpToEven = (value) => {
  const rounded = Math.max(2, Math.ceil(value));
  return rounded % 2 === 0 ? rounded : rounded + 1;
};
var syncCanvasSize = (canvas, width, height, density) => {
  const scaledWidth = roundUpToEven(width * density);
  const scaledHeight = roundUpToEven(height * density);
  if (canvas.width !== scaledWidth) {
    canvas.width = scaledWidth;
  }
  if (canvas.height !== scaledHeight) {
    canvas.height = scaledHeight;
  }
};
var getCursorForElement = (element) => {
  let current = element;
  while (current) {
    const { cursor } = window.getComputedStyle(current);
    if (cursor !== "auto") {
      return cursor;
    }
    current = current.parentElement;
  }
  return "auto";
};
var downloadBlob = (blob, filename) => {
  const url = URL.createObjectURL(blob);
  const anchor = document.createElement("a");
  anchor.href = url;
  anchor.download = filename;
  document.body.appendChild(anchor);
  anchor.click();
  document.body.removeChild(anchor);
  URL.revokeObjectURL(url);
};
var getJsonFilename = (filename) => {
  return filename.replace(/\.[^.]+$/, "") + ".json";
};
var logCaptureError = (message, err) => {
  console.error(message, err instanceof Error ? err.message : String(err));
};
var addPaintFrameToRecording = (recording, source) => {
  const elapsedInSeconds = Math.max(0, (performance.now() - recording.startedAt) / 1000);
  const timestampInSeconds = recording.lastTimestampInSeconds === null ? 0 : elapsedInSeconds;
  const durationInSeconds = recording.lastTimestampInSeconds === null ? fallbackFrameDurationInSeconds : Math.max(fallbackFrameDurationInSeconds, elapsedInSeconds - recording.lastTimestampInSeconds);
  const keyFrame = recording.lastTimestampInSeconds === null;
  recording.lastTimestampInSeconds = timestampInSeconds;
  recording.lastFramePromise = recording.lastFramePromise.then(async () => {
    await source.add(timestampInSeconds, durationInSeconds, { keyFrame });
    recording.frameCount++;
  });
};
var finalizeRecording = async (recording, filename) => {
  addPaintFrameToRecording(recording, recording.source);
  await recording.lastFramePromise;
  if (recording.frameCount === 0) {
    throw new Error("No frames were added to the Studio canvas recording.");
  }
  recording.source.close();
  await recording.output.finalize();
  if (!recording.target.buffer) {
    throw new Error("Mediabunny did not return an output buffer.");
  }
  downloadBlob(new Blob([recording.target.buffer], { type: "video/webm" }), filename);
  downloadBlob(new Blob([
    JSON.stringify({
      startedAt: recording.startedAt,
      endedAt: performance.now(),
      captureMetadata: recording.captureMetadata,
      mouseMovements: recording.mouseMovements
    }, null, 2)
  ], { type: "application/json" }), getJsonFilename(filename));
};
var HtmlInCanvasCapture = forwardRef(({ children, density, filename }, ref) => {
  if (!Number.isFinite(density) || density <= 0) {
    throw new Error("HTML-in-canvas capture density must be greater than 0.");
  }
  const isSupported = useMemo(() => isHtmlInCanvasAvailable(), []);
  const canvasRef = useRef(null);
  const contentRef = useRef(null);
  const recordingRef = useRef(null);
  const recordingActionRef = useRef(Promise.resolve());
  const requestPaint = useCallback(() => {
    const canvas = canvasRef.current;
    if (typeof canvas?.requestPaint !== "function") {
      return;
    }
    canvas.requestPaint();
  }, []);
  const startRecording = useCallback(async () => {
    const canvas = canvasRef.current;
    if (!canvas || recordingRef.current) {
      return;
    }
    const { BufferTarget, CanvasSource, Output, WebMOutputFormat } = await import("mediabunny");
    const target = new BufferTarget;
    const output = new Output({
      format: new WebMOutputFormat,
      target
    });
    const source = new CanvasSource(canvas, {
      codec: "vp9",
      bitrate: recordingVideoBitrate,
      latencyMode: "realtime",
      keyFrameInterval: recordingKeyFrameIntervalInSeconds
    });
    output.addVideoTrack(source);
    await output.start();
    recordingRef.current = {
      output,
      target,
      source,
      startedAt: performance.now(),
      mouseMovements: [],
      lastTimestampInSeconds: null,
      lastFramePromise: Promise.resolve(),
      frameCount: 0,
      captureMetadata: null,
      isFinalizing: false
    };
    requestPaint();
  }, [requestPaint]);
  const stopRecording = useCallback(async () => {
    const recording = recordingRef.current;
    if (!recording || recording.isFinalizing) {
      return;
    }
    recording.isFinalizing = true;
    try {
      await finalizeRecording(recording, filename);
    } catch (err) {
      logCaptureError("Could not finalize Studio canvas recording:", err);
      await recording.output.cancel().catch((cancelErr) => {
        logCaptureError("Could not cancel Studio canvas recording:", cancelErr);
      });
    } finally {
      recordingRef.current = null;
    }
  }, [filename]);
  const toggleRecording = useCallback(async () => {
    recordingActionRef.current = recordingActionRef.current.then(async () => {
      if (recordingRef.current) {
        await stopRecording();
        return;
      }
      await startRecording();
    });
    await recordingActionRef.current;
  }, [startRecording, stopRecording]);
  useImperativeHandle(ref, () => ({
    toggleRecording,
    startRecording,
    stopRecording
  }), [startRecording, stopRecording, toggleRecording]);
  const drawCurrentPaint = useCallback(() => {
    const canvas = canvasRef.current;
    const content = contentRef.current;
    if (!canvas || !content) {
      return;
    }
    const context = canvas.getContext("2d");
    if (!context || typeof context.drawElementImage !== "function") {
      throw new Error("drawElementImage() is not available.");
    }
    const rect = content.getBoundingClientRect();
    if (rect.width <= 0 || rect.height <= 0) {
      return;
    }
    syncCanvasSize(canvas, rect.width, rect.height, density);
    resetCanvas(context, canvas);
    context.scale(density, density);
    context.drawElementImage(content, 0, 0, rect.width, rect.height);
    const recording = recordingRef.current;
    if (recording && !recording.isFinalizing) {
      recording.captureMetadata = {
        density,
        contentRect: {
          left: rect.left,
          top: rect.top,
          width: rect.width,
          height: rect.height
        },
        canvasSize: {
          width: canvas.width,
          height: canvas.height
        },
        viewport: {
          width: window.innerWidth,
          height: window.innerHeight,
          scrollX: window.scrollX,
          scrollY: window.scrollY
        }
      };
      addPaintFrameToRecording(recording, recording.source);
    }
  }, [density]);
  useEffect(() => {
    const onMouseMove = (event) => {
      const recording = recordingRef.current;
      if (!recording || recording.isFinalizing) {
        return;
      }
      const rect = contentRef.current?.getBoundingClientRect();
      recording.mouseMovements.push({
        timeInSeconds: (performance.now() - recording.startedAt) / 1000,
        clientX: event.clientX,
        clientY: event.clientY,
        pageX: event.pageX,
        pageY: event.pageY,
        canvasX: rect ? (event.clientX - rect.left) * density : null,
        canvasY: rect ? (event.clientY - rect.top) * density : null,
        cursor: getCursorForElement(document.elementFromPoint(event.clientX, event.clientY))
      });
    };
    window.addEventListener("pointermove", onMouseMove);
    return () => {
      window.removeEventListener("pointermove", onMouseMove);
    };
  }, [density]);
  useEffect(() => {
    if (!isSupported) {
      return;
    }
    const canvas = canvasRef.current;
    if (!canvas) {
      return;
    }
    canvas.setAttribute("layoutsubtree", "");
    canvas.layoutSubtree = true;
    const onPaint = () => {
      try {
        drawCurrentPaint();
      } catch (err) {
        logCaptureError("Could not capture Studio canvas paint:", err);
      }
    };
    canvas.addEventListener("paint", onPaint);
    const frame = requestAnimationFrame(requestPaint);
    return () => {
      cancelAnimationFrame(frame);
      canvas.removeEventListener("paint", onPaint);
    };
  }, [drawCurrentPaint, isSupported, requestPaint]);
  useEffect(() => {
    if (!isSupported) {
      return;
    }
    const content = contentRef.current;
    const canvas = canvasRef.current;
    if (!content || !canvas) {
      return;
    }
    const resizeObserver = new ResizeObserver(([entry]) => {
      const { width, height } = entry.contentRect;
      syncCanvasSize(canvas, width, height, density);
      requestPaint();
    });
    resizeObserver.observe(content);
    return () => {
      resizeObserver.disconnect();
    };
  }, [density, isSupported, requestPaint]);
  useEffect(() => {
    return () => {
      const recording = recordingRef.current;
      if (!recording || recording.isFinalizing) {
        return;
      }
      recording.isFinalizing = true;
      recording.output.cancel().catch((err) => {
        logCaptureError("Could not cancel Studio canvas recording:", err);
      });
      recordingRef.current = null;
    };
  }, []);
  if (!isSupported) {
    return children;
  }
  return /* @__PURE__ */ jsx("canvas", {
    ref: canvasRef,
    style: canvasStyle,
    children: /* @__PURE__ */ jsx("div", {
      ref: contentRef,
      style: contentStyle,
      children
    })
  });
});
var withHtmlInCanvasCapture = (Component) => {
  return forwardRef(({ density, filename, ...props }, ref) => {
    return /* @__PURE__ */ jsx(HtmlInCanvasCapture, {
      ref,
      density,
      filename,
      children: /* @__PURE__ */ jsx(Component, {
        ...props
      })
    });
  });
};
export {
  withHtmlInCanvasCapture,
  isHtmlInCanvasAvailable,
  HtmlInCanvasCapture
};
