// @remotion/studio-server/dist/codemods/recast-mods.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.applyCodemod = void 0;
const apply_visual_control_1 = require("./apply-visual-control");
const applyCodemod = ({ file, codeMod, }) => {
    const changesMade = [];
    if (codeMod.type === 'apply-visual-control') {
        return (0, apply_visual_control_1.applyVisualControl)({ file, transformation: codeMod, changesMade });
    }
    const body = file.program.body.map((node) => {
        return mapAll(node, codeMod, changesMade, null);
    });
    return {
        newAst: {
            ...file,
            program: {
                ...file.program,
                body,
            },
        },
        changesMade,
    };
};
exports.applyCodemod = applyCodemod;
const mapAll = (node, transformation, changesMade, parentFolderName) => {
    if (isRecognizedType(node)) {
        return mapRecognizedType(node, transformation, changesMade, parentFolderName);
    }
    return node;
};
const isRecognizedType = (t) => {
    return (t.type === 'ArrowFunctionExpression' ||
        t.type === 'FunctionExpression' ||
        t.type === 'JSXFragment' ||
        t.type === 'JSXElement' ||
        t.type === 'BlockStatement' ||
        t.type === 'ReturnStatement' ||
        t.type === 'VariableDeclaration' ||
        t.type === 'FunctionDeclaration' ||
        t.type === 'ExportNamedDeclaration' ||
        t.type === 'ExportDefaultDeclaration');
};
const mapVariableDeclarator = (variableDeclarator, transformation, changesMade, parentFolderName) => {
    return {
        ...variableDeclarator,
        init: variableDeclarator.init
            ? mapAll(variableDeclarator.init, transformation, changesMade, parentFolderName)
            : variableDeclarator.init,
    };
};
const mapBlockStatement = (blockStatement, transformation, changesMade, parentFolderName) => {
    return {
        ...blockStatement,
        body: blockStatement.body.map((a) => {
            return mapAll(a, transformation, changesMade, parentFolderName);
        }),
    };
};
const mapReturnStatement = (statement, transformation, changesMade, parentFolderName) => {
    if (!statement.argument) {
        return statement;
    }
    const replacement = transformLoneJsxElement(statement.argument, transformation, changesMade, parentFolderName);
    if (replacement !== null) {
        return { ...statement, argument: replacement };
    }
    return {
        ...statement,
        argument: mapAll(statement.argument, transformation, changesMade, parentFolderName),
    };
};
const nullLiteral = () => ({ type: 'NullLiteral' });
// When a node was originally parenthesized (e.g. the `<JSX/>` inside
// `return (<JSX/>)`), Babel records `extra.parenthesized` on it. If we move
// such a node into a JSXFragment without clearing that hint, recast prints
// stray `(` / `)` characters around it as JSX text. Clear the hint so the
// node prints cleanly in its new context.
const stripParenthesizedExtra = (node) => {
    if (!node.extra) {
        return node;
    }
    const { parenthesized: _p, parenStart: _ps, ...rest } = node.extra;
    return { ...node, extra: rest };
};
const wrapInJsxFragment = (children) => ({
    type: 'JSXFragment',
    openingFragment: { type: 'JSXOpeningFragment' },
    closingFragment: { type: 'JSXClosingFragment' },
    children: children.map((child) => {
        if (child.type === 'JSXElement' || child.type === 'JSXFragment') {
            return stripParenthesizedExtra(child);
        }
        return child;
    }),
});
const isJsxExpression = (child) => {
    return child.type === 'JSXElement' || child.type === 'JSXFragment';
};
const isMeaningfulJsxChild = (child) => {
    return child.type !== 'JSXText' || child.value.trim() !== '';
};
// When a <Composition> JSX element appears in a position where it cannot
// simply be removed from a parent's children list (e.g. as the sole return
// value of a wrapper component or as the concise body of an arrow function),
// we still want delete/rename/duplicate codemods to work. This helper detects
// that case and produces a structurally-valid replacement expression.
const transformLoneJsxElement = (expression, transformation, changesMade, parentFolderName) => {
    if (expression.type !== 'JSXElement') {
        return null;
    }
    const compId = getCompositionIdFromJSXElement(expression);
    const folderName = getFolderNameFromJSXElement(expression);
    if (compId === null) {
        const isFolderMatch = folderName !== null &&
            ((transformation.type === 'delete-folder' &&
                folderName === transformation.folderName &&
                parentFolderName === transformation.parentName) ||
                (transformation.type === 'rename-folder' &&
                    folderName === transformation.folderName &&
                    parentFolderName === transformation.parentName));
        if (!isFolderMatch) {
            return null;
        }
    }
    const isMatch = (transformation.type === 'delete-composition' &&
        compId === transformation.idToDelete) ||
        (transformation.type === 'rename-composition' &&
            compId === transformation.idToRename) ||
        (transformation.type === 'duplicate-composition' &&
            compId === transformation.idToDuplicate) ||
        (transformation.type === 'delete-folder' &&
            folderName === transformation.folderName &&
            parentFolderName === transformation.parentName) ||
        (transformation.type === 'rename-folder' &&
            folderName === transformation.folderName &&
            parentFolderName === transformation.parentName);
    if (!isMatch) {
        return null;
    }
    const transformed = mapJsxChild(expression, transformation, changesMade, parentFolderName);
    const meaningful = transformed.filter(isMeaningfulJsxChild);
    if (meaningful.length === 0) {
        return nullLiteral();
    }
    if (meaningful.length === 1 && isJsxExpression(meaningful[0])) {
        return meaningful[0];
    }
    return wrapInJsxFragment(transformed);
};
const mapJsxElementOrFragment = (jsxFragment, transformation, changesMade, parentFolderName) => {
    return {
        ...jsxFragment,
        children: jsxFragment.children
            .map((c) => {
            if (c.type !== 'JSXElement') {
                return c;
            }
            return mapJsxChild(c, transformation, changesMade, parentFolderName);
        })
            .flat(1),
    };
};
const getChildFolderParentName = ({ folderName, parentFolderName, }) => {
    return [parentFolderName, folderName].filter(Boolean).join('/');
};
const mapJsxChild = (c, transformation, changesMade, parentFolderName) => {
    const compId = getCompositionIdFromJSXElement(c);
    const folderName = getFolderNameFromJSXElement(c);
    if (transformation === null) {
        return [c];
    }
    if (transformation.type === 'duplicate-composition' &&
        compId === transformation.idToDuplicate) {
        return [
            c,
            changeComposition({
                jsxElement: c,
                newCompositionId: transformation.newId,
                newCompositionFps: transformation.newFps,
                newCompositionDurationInFrames: transformation.newDurationInFrames,
                newCompositionHeight: transformation.newHeight,
                newCompositionWidth: transformation.newWidth,
                newTagToUse: transformation.tag,
                changesMade,
            }),
        ];
    }
    if (transformation.type === 'rename-composition' &&
        compId === transformation.idToRename) {
        return [
            changeComposition({
                jsxElement: c,
                newCompositionId: transformation.newId,
                newCompositionFps: null,
                newCompositionDurationInFrames: null,
                newCompositionHeight: null,
                newCompositionWidth: null,
                changesMade,
                newTagToUse: null,
            }),
        ];
    }
    if (transformation.type === 'delete-composition' &&
        compId === transformation.idToDelete) {
        changesMade.push({
            description: 'Deleted composition',
        });
        return [];
    }
    if (transformation.type === 'rename-folder' &&
        folderName === transformation.folderName &&
        parentFolderName === transformation.parentName) {
        return [
            changeFolderName({
                jsxElement: c,
                newFolderName: transformation.newName,
                changesMade,
            }),
        ];
    }
    if (transformation.type === 'delete-folder' &&
        folderName === transformation.folderName &&
        parentFolderName === transformation.parentName) {
        changesMade.push({
            description: 'Deleted folder',
        });
        return c.children;
    }
    const childParentFolderName = folderName
        ? getChildFolderParentName({ folderName, parentFolderName })
        : parentFolderName;
    return [mapAll(c, transformation, changesMade, childParentFolderName)];
};
const mapRecognizedType = (expression, transformation, changesMade, parentFolderName) => {
    if (expression.type === 'JSXFragment' || expression.type === 'JSXElement') {
        return mapJsxElementOrFragment(expression, transformation, changesMade, parentFolderName);
    }
    if (expression.type === 'ArrowFunctionExpression' ||
        expression.type === 'FunctionExpression') {
        if (expression.type === 'ArrowFunctionExpression' &&
            expression.body.type === 'JSXElement') {
            const replacement = transformLoneJsxElement(expression.body, transformation, changesMade, parentFolderName);
            if (replacement !== null) {
                return {
                    ...expression,
                    body: replacement,
                };
            }
        }
        return {
            ...expression,
            body: mapAll(expression.body, transformation, changesMade, parentFolderName),
        };
    }
    if (expression.type === 'VariableDeclaration') {
        const declarations = expression.declarations.map((d) => {
            return mapVariableDeclarator(d, transformation, changesMade, parentFolderName);
        });
        return { ...expression, declarations };
    }
    if (expression.type === 'FunctionDeclaration') {
        return {
            ...expression,
            body: mapBlockStatement(expression.body, transformation, changesMade, parentFolderName),
        };
    }
    if (expression.type === 'ExportNamedDeclaration' ||
        expression.type === 'ExportDefaultDeclaration') {
        if (!expression.declaration) {
            return expression;
        }
        return {
            ...expression,
            declaration: mapAll(expression.declaration, transformation, changesMade, parentFolderName),
        };
    }
    if (expression.type === 'ReturnStatement') {
        return mapReturnStatement(expression, transformation, changesMade, parentFolderName);
    }
    if (expression.type === 'BlockStatement') {
        return mapBlockStatement(expression, transformation, changesMade, parentFolderName);
    }
    return expression;
};
const getCompositionIdFromJSXElement = (jsxElement) => {
    if (jsxElement.type !== 'JSXElement') {
        return null;
    }
    const { openingElement } = jsxElement;
    const { name } = openingElement;
    if (name.type !== 'JSXIdentifier') {
        return null;
    }
    if (name.name !== 'Composition' && name.name !== 'Still') {
        return null;
    }
    const id = openingElement.attributes
        .map((attribute) => {
        if (attribute.type === 'JSXSpreadAttribute') {
            return null;
        }
        if (attribute.name.type === 'JSXNamespacedName') {
            return null;
        }
        if (attribute.name.name !== 'id') {
            return null;
        }
        if (!attribute.value) {
            return null;
        }
        if (attribute.value.type === 'StringLiteral') {
            return attribute.value.value;
        }
        if (attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'StringLiteral') {
            return attribute.value.expression.value;
        }
        return null;
    })
        .filter(Boolean);
    return id[0];
};
const getFolderNameFromJSXElement = (jsxElement) => {
    if (jsxElement.type !== 'JSXElement') {
        return null;
    }
    const { openingElement } = jsxElement;
    const { name } = openingElement;
    if (name.type !== 'JSXIdentifier') {
        return null;
    }
    if (name.name !== 'Folder') {
        return null;
    }
    const folderName = openingElement.attributes
        .map((attribute) => {
        if (attribute.type === 'JSXSpreadAttribute') {
            return null;
        }
        if (attribute.name.type === 'JSXNamespacedName') {
            return null;
        }
        if (attribute.name.name !== 'name') {
            return null;
        }
        if (!attribute.value) {
            return null;
        }
        if (attribute.value.type === 'StringLiteral') {
            return attribute.value.value;
        }
        if (attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'StringLiteral') {
            return attribute.value.expression.value;
        }
        return null;
    })
        .filter(Boolean);
    return folderName[0];
};
const changeFolderName = ({ jsxElement, newFolderName, changesMade, }) => {
    const { openingElement } = jsxElement;
    const { name } = openingElement;
    if (name.type !== 'JSXIdentifier') {
        return jsxElement;
    }
    if (name.name !== 'Folder') {
        return jsxElement;
    }
    const attributes = openingElement.attributes.map((attribute) => {
        if (attribute.type === 'JSXSpreadAttribute') {
            return attribute;
        }
        if (attribute.name.type === 'JSXNamespacedName') {
            return attribute;
        }
        if (attribute.name.name === 'name' &&
            attribute.value &&
            attribute.value.type === 'StringLiteral') {
            changesMade.push({
                description: 'Replaced folder name',
            });
            return {
                ...attribute,
                value: { ...attribute.value, value: newFolderName },
            };
        }
        if (attribute.name.name === 'name' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'StringLiteral') {
            changesMade.push({
                description: 'Replaced folder name',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newFolderName,
                    },
                },
            };
        }
        return attribute;
    });
    return {
        ...jsxElement,
        openingElement: {
            ...jsxElement.openingElement,
            attributes,
        },
    };
};
const changeComposition = ({ jsxElement, newCompositionId, newCompositionFps, newCompositionDurationInFrames, newCompositionHeight, newCompositionWidth, changesMade, newTagToUse, }) => {
    const { openingElement } = jsxElement;
    const { name } = openingElement;
    if (name.type !== 'JSXIdentifier') {
        return jsxElement;
    }
    if (name.name !== 'Composition' && name.name !== 'Still') {
        return jsxElement;
    }
    const attributes = openingElement.attributes
        .map((attribute) => {
        if (attribute.type === 'JSXSpreadAttribute') {
            return attribute;
        }
        if (attribute.name.type === 'JSXNamespacedName') {
            return attribute;
        }
        if (attribute.name.name === 'fps' && newTagToUse === 'Still') {
            changesMade.push({ description: 'Removed fps attribute' });
            return null;
        }
        if (attribute.name.name === 'durationInFrames' &&
            newTagToUse === 'Still') {
            changesMade.push({ description: 'Removed durationInFrames' });
            return null;
        }
        // id="one"
        if (attribute.name.name === 'id' &&
            attribute.value &&
            attribute.value.type === 'StringLiteral') {
            changesMade.push({
                description: 'Replaced composition id',
            });
            return {
                ...attribute,
                value: { ...attribute.value, value: newCompositionId },
            };
        }
        // id={"one"}
        if (attribute.name.name === 'id' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'StringLiteral') {
            changesMade.push({
                description: 'Replaced composition id',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newCompositionId,
                    },
                },
            };
        }
        if (attribute.name.name === 'fps' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'NumericLiteral' &&
            newCompositionFps !== null) {
            changesMade.push({
                description: 'Replaced FPS',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newCompositionFps,
                    },
                },
            };
        }
        if (attribute.name.name === 'durationInFrames' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'NumericLiteral' &&
            newCompositionDurationInFrames !== null) {
            changesMade.push({
                description: 'Replaced durationInFrames',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newCompositionDurationInFrames,
                    },
                },
            };
        }
        if (attribute.name.name === 'width' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'NumericLiteral' &&
            newCompositionWidth !== null) {
            changesMade.push({
                description: 'Replaced width',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newCompositionWidth,
                    },
                },
            };
        }
        if (attribute.name.name === 'height' &&
            attribute.value &&
            attribute.value.type === 'JSXExpressionContainer' &&
            attribute.value.expression.type === 'NumericLiteral' &&
            newCompositionHeight !== null) {
            changesMade.push({
                description: 'Replaced height',
            });
            return {
                ...attribute,
                value: {
                    ...attribute.value,
                    expression: {
                        ...attribute.value.expression,
                        value: newCompositionHeight,
                    },
                },
            };
        }
        return attribute;
    })
        .filter(Boolean);
    const newName = newTagToUse !== null && newTagToUse !== void 0 ? newTagToUse : name.name;
    if (newName !== name.name) {
        changesMade.push({
            description: `Changed tag`,
        });
    }
    return {
        ...jsxElement,
        openingElement: {
            ...jsxElement.openingElement,
            name: {
                ...name,
                name: newName,
            },
            attributes,
        },
    };
};
